### PR TITLE
[BFD-858] SPIKE: Aurora Upgrade Postgres to next available Major Version

### PR DESCRIPTION
**Pull request by mackec1**
### Change Details

BFD-858 was a spike to investigate aAurora database engin upgrade from 11.6 to 12.4. While the SPIKE document was checked into Confluence, s pseudo-copy of the document as also added to the ccs-ops-misc folder.

### Acceptance Validation

The README markdown provides information and best practices for upgrading an Aurora database engine.

### Feedback Requested

Missing ops steps and/or smoke-testing details.

### External References

- [BFD-858](https://jira.cms.gov/browse/BFD-858)

### Security Implications

- [ X] new software dependencies

End result of a db engine upgrade by definition introduces new software dependencies; however, BFD services API use JDBC 4.2 which remains unchanged.

- [ ] altered security controls

- [ ] new data stored or transmitted

- [ ] security checklist is completed for this change

- [ ] requires more information or team discussion to evaluate security implications

### CODE REVIEW

File: ops/ccs-ops-misc/aurora-db-engine-upgrade/README.md [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+AWS provides both the tools and best practices to upgrade an RDS Postgres database engine to next available major/minor version; the steps to safely upgrade the BFD database will be outline here. The most important planning component will be the timing of when to perform the database clone of the current BFD database. Once, a clone database has been created, the actual upgrade steps can begin.
+
+Note: The following steps/examples apply to all BFD Aurora database(s):
+
+-   bfd-prod
+-   bfd-prod-sbx
+-   bfd-test
+
+### Plan the Postgres db Engine Upgrade Path
```
Comment (jzulim) [inline ops/ccs-ops-misc/aurora-db-engine-upgrade/README.md:113]: @mackec1 It seems this doc might be duplicated? I'm seeing repeated sections.

**Review by jzulim — APPROVED** [review]
LGTM

**Comment by keithdadkins** [comment]
I see that we upgraded our db instances to 11.9 from 11.6. You should make a note that anytime we make infrastructure changes out-of-band, we need to update our infrastructure-as-code (terraform in this case) to reflect those changes.

IE. We need to update the following 3 files to reflect the current db version (line 24 in main.tf):
- ops/terraform/env/prod/stateful/main.tf
- ops/terraform/env/prod-sbx/stateful/main.tf
- ops/terraform/env/test/stateful/main.tf

```terraform
...
  aurora_config = {
    instance_class = "db.r5.12xlarge"

    # With aurora you do not designate primary and replicas. Instead, you simply add RDS Instances and
    # Aurora manages the replication. So if you want 1 writer and 3 readers, you set cluster_nodes to 4
    cluster_nodes  = 4
    engine_version = "11.6" # <---------CHANGE THIS
    param_version  = "aurora-postgresql11"
  }
```
Terraform does not upgrade the cluster. So when the versions are mismatched, it will try to **replace** it instead. This of course would be bad.

**Comment by keithdadkins** [comment]
Even though our data is currently loaded over the weekend (starting on Fridays), that may not always be the case so we should get in the habit of creating a database snapshot prior to upgrading our live production cluster. This is to ensure we have the latest point-in-time to restore from.

**Review by keithdadkins — CHANGES_REQUESTED** [review]
Please see my PR comments.

**Review by JeremyIR — APPROVED** [review]
Easy to follow and will be nice to have at hand when needed. LGTM

**Comment by jzulim** [comment]
> Even though our data is currently loaded over the weekend (starting on Fridays), that may not always be the case so we should get in the habit of creating a database snapshot prior to upgrading our live production cluster. This is to ensure we have the latest point-in-time to restore from.

100 - When DC GEO begins investing, we'll dream / load data all the time.

**Comment by mackec1** [comment]
no longer needed; Keith has PR(s) that do the same and have been tested in real life.

**Pull request closed without merging**